lib/gauge_polyakov_loop.cu

```diff
@@ -0,0 +1,276 @@
+
+      std::unique_ptr<GaugeField> product_field = std::make_unique<cudaGaugeField>(gParam);
+      GaugeField& product_field_ref = reinterpret_cast<GaugeField&>(*product_field.get());
+
+      // Create the field we reduce into
+      x[3] = comm_dim(3);
+      gParam.x = x;
+      gParam.create = QUDA_NULL_FIELD_CREATE;
+      condensed_field = std::make_unique<cudaGaugeField>(gParam);
+      GaugeField& condensed_field_ref = reinterpret_cast<GaugeField&>(*condensed_field.get());
+      profile.TPSTOP(QUDA_PROFILE_INIT);
+
+      profile.TPSTART(QUDA_PROFILE_COMPUTE);
+
+      // Compute the rank-local product of loops in the temporal direction...
+      instantiate<GaugePolyakovLoopProduct, ReconstructNo12>(u, product_field_ref);
+      // ...and inject it into my own accumulation field
+      gaugeInsertTimeslice(condensed_field_ref, product_field_ref, comm_coord(3));
+      profile.TPSTOP(QUDA_PROFILE_COMPUTE);
+
+      // we can skip comms if we're only doing a partition test
+      if (comm_dim(dir) > 1) {
+        // Send/gather other data. For the time being this does host staging, but we could
+        // optimize this to us device-aware MPI.
+        profile.TPSTART(QUDA_PROFILE_INIT);
+        auto bytes = product_field_ref.TotalBytes();
+        void* v_double_buffer[2];
+        MsgHandle *recv_handle[2];
+        MsgHandle *send_handle[2];
+        for (int i = 0; i < 2; i++) {
+          v_double_buffer[i] = pinned_malloc(bytes);
+          // receive from in front of me, send behind
+          recv_handle[i] = comm_declare_receive_relative(v_double_buffer[i], dir, 1, bytes);
+          send_handle[i] = comm_declare_send_relative(v_double_buffer[i], dir, -1, bytes);
+        }
+        int SEND_BUFFER = 0;
+        int RECV_BUFFER = 1;
+        profile.TPSTOP(QUDA_PROFILE_INIT);
+
+        // prepare first send
+        profile.TPSTART(QUDA_PROFILE_D2H);
+        product_field_ref.copy_to_buffer(v_double_buffer[SEND_BUFFER]);
+        profile.TPSTOP(QUDA_PROFILE_D2H);
+
+        // kick off
+        for (int t = 1; t < comm_dim(3); t++) {
+          profile.TPSTART(QUDA_PROFILE_COMMS);
+          // post a receive from rank behind me
+          comm_start(recv_handle[RECV_BUFFER]);
+          comm_start(send_handle[SEND_BUFFER]);
+
+          // wait
+          comm_wait(recv_handle[RECV_BUFFER]);
+          comm_wait(send_handle[SEND_BUFFER]);
+          profile.TPSTOP(QUDA_PROFILE_COMMS);
+
+          // copy the received buffer into our staging field
+          profile.TPSTART(QUDA_PROFILE_H2D);
+          product_field_ref.copy_from_buffer(v_double_buffer[RECV_BUFFER]);
+          profile.TPSTOP(QUDA_PROFILE_H2D);
+
+          // insert
+          profile.TPSTART(QUDA_PROFILE_COMPUTE);
+          gaugeInsertTimeslice(condensed_field_ref, product_field_ref, (comm_coord(3) + t) % comm_dim(3));
+          profile.TPSTOP(QUDA_PROFILE_COMPUTE);
+
+          // swap buffers
+          SEND_BUFFER ^= 1;
```

Comment:
I think this is unnecessary optimization that only serves to obfuscate the meaning here.

Reply:
Me? Never! (Jokes aside, replaced by a `std::swap`)